server/lomas_server/utils/collection_models.py

```diff
@@ -88,11 +89,156 @@ class DatasetsCollection(BaseModel):
     ]
 
 
+class ColumnMetadata(BaseModel):
+    """Base model for column metadata"""
+
+    private_id: bool = False
+    nullable: bool = False
+    # TODO create validator and test these, see issue #323
+    max_partition_length: Optional[Annotated[int, Field(gt=0)]] = None
+    max_influenced_partitions: Optional[Annotated[int, Field(gt=0)]] = None
+    max_partition_contributions: Optional[Annotated[int, Field(gt=0)]] = None
+
+
+class StrMetadata(ColumnMetadata):
+    """Model for string metadata"""
+
+    type: Literal["string"]
+
+
+class CategoricalColumnMetadata(ColumnMetadata):
+    """Model for categorical column metadata"""
+
+    @model_validator(mode="after")
+    def validate_categories(self):
+        """Makes sure number of categories matches cardinality."""
+        if len(self.categories) != self.cardinality:
+            raise ValueError(
+                "Number of categories should be equal to cardinality."
+            )
+        return self
+
+
+class StrCategoricalMetadata(CategoricalColumnMetadata):
+    """Model for categorical string metadata"""
+
+    type: Literal["string"]
+    cardinality: int
+    categories: List[str]
+
+
+class Precision(IntEnum):
+    """Precision of integer and float data"""
+
+    SINGLE = 32
+    DOUBLE = 64
+
+
+class BoundedColumnMetadata(ColumnMetadata):
+    """Model for columns with bounded data"""
+
+    @model_validator(mode="after")
+    def validate_bounds(self):
+        """Validates column bounds."""
+        if (
+            self.lower is not None
+            and self.upper is not None
+            and self.lower > self.upper
+        ):
+            raise ValueError("Lower bound cannot be larger than upper bound.")
+
+        return self
+
+
+class IntMetadata(BoundedColumnMetadata):
+    """Model for integer column metadata"""
+
+    type: Literal["int"]
+    precision: Precision
+    lower: int
+    upper: int
+
+
+class IntCategoricalMetadata(CategoricalColumnMetadata):
+    """Model for integer categorical column metadata"""
+
+    type: Literal["int"]
+    precision: Precision
+    cardinality: int
+    categories: List[int]
+
+
+class FloatMetadata(BoundedColumnMetadata):
+    """Model for float column metadata"""
+
+    type: Literal["float"]
+    precision: Precision
+    lower: float
+    upper: float
+
+
+class BooleanMetadata(ColumnMetadata):
+    """Model for boolean column metadata"""
+
+    type: Literal["boolean"]
+
+
+class DatetimeMetadata(BoundedColumnMetadata):
+    """Model for datetime column metadata"""
+
+    type: Literal["datetime"]  # TODO make these constants, see issue #268
+    lower: datetime
+    upper: datetime
```

Comment:
will this convert the data to datetime automatically ?

Reply:
like no need for date= datetime.strptime(date_in_str_in_metadata, DEFAULT_DATE_FORMAT)

Reply:
Yes